File: web/siteplan/src/model/PZB.ts
/**
 * Copyright (c) 2021 DB Netz AG and others.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v20.html
 */
import { checkInstance } from '@/util/ObjectExtension'
import { Label } from './Label'
import { defaultPositionObj, Position } from './Position'
import { defaultRouteLocationObj, RouteObject } from './RouteObject'
import SiteplanObject, { defaultObjectColorObj } from './SiteplanObject'

export enum PZBType {
    GM = 'GM',
    GUE_GSA = 'GUE_GSA',
    GUE_GSE = 'GUE_GSE'
}

export enum PZBElement {
    None = 'None',
    F500Hz = '500Hz',
    F1000Hz = '1000Hz',
    F2000Hz = '2000Hz',
    F1000Hz2000Hz = '1000Hz2000Hz',
}

export enum PZBEffectivity {
    None = 'None',
    Signal = 'Signal',
    Route = 'Route',
    Always = 'Always'
}

export interface PZB extends RouteObject, SiteplanObject
{
    guid: string
    position: Position
    type: PZBType
    element: PZBElement
    rightSide: boolean
    effectivity: PZBEffectivity
}

export function defaultPZBObj (): PZB {
  return {
    guid: '0',
    position: defaultPositionObj(),
    type: PZBType.GM,
    element: PZBElement.F1000Hz,
    rightSide: true,
    effectivity: PZBEffectivity.Always,
    objectColors: [defaultObjectColorObj()],
    routeLocations: [defaultRouteLocationObj()]
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function instanceofPZB (object: any): boolean {
  return checkInstance(object, defaultPZBObj())
}

/**
     * Returns a label for a PZB
     * @param pzb the PZB
     * @returns a label for the PZB
     */
export function getPZBLabel (pzb: PZB): Label {
  let text = ''
  switch (pzb.type) {
    case PZBType.GUE_GSA:
      text = 'GSA'
      break
    case PZBType.GUE_GSE:
      text = 'GSE'
      break
    case PZBType.GM:
      switch (pzb.element) {
        case PZBElement.F1000Hz:
          text = '1'
          break
        case PZBElement.F2000Hz:
          text = '2'
          break
        case PZBElement.F500Hz:
          text = '0,5'
          break
        case PZBElement.F1000Hz2000Hz:
          text = '1/2'
          break
        default:
      }

      break
    default:
  }

  return {
    text,
    orientationInverted: true
  } as Label
}

File: java/bundles/org.eclipse.set.model.siteplan/src/org/eclipse/set/model/siteplan/PZB.java
/**
 * Copyright (c) 2021 DB Netz AG and others.
 * 
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v20.html
 */
package org.eclipse.set.model.siteplan;

/**
 * <!-- begin-user-doc --> A representation of the model object
 * '<em><b>PZB</b></em>'. <!-- end-user-doc -->
 *
 * <p>
 * The following features are supported:
 * </p>
 * <ul>
 *   <li>{@link org.eclipse.set.model.siteplan.PZB#getType <em>Type</em>}</li>
 *   <li>{@link org.eclipse.set.model.siteplan.PZB#getElement <em>Element</em>}</li>
 *   <li>{@link org.eclipse.set.model.siteplan.PZB#isRightSide <em>Right Side</em>}</li>
 *   <li>{@link org.eclipse.set.model.siteplan.PZB#getEffectivity <em>Effectivity</em>}</li>
 * </ul>
 *
 * @see org.eclipse.set.model.siteplan.SiteplanPackage#getPZB()
 * @model
 * @generated
 */
public interface PZB extends RouteObject, PositionedObject {
	/**
	 * Returns the value of the '<em><b>Type</b></em>' attribute.
	 * The literals are from the enumeration {@link org.eclipse.set.model.siteplan.PZBType}.
	 * <!-- begin-user-doc --> <!-- end-user-doc -->
	 * @return the value of the '<em>Type</em>' attribute.
	 * @see org.eclipse.set.model.siteplan.PZBType
	 * @see #setType(PZBType)
	 * @see org.eclipse.set.model.siteplan.SiteplanPackage#getPZB_Type()
	 * @model
	 * @generated
	 */
	PZBType getType();

	/**
	 * Sets the value of the '{@link org.eclipse.set.model.siteplan.PZB#getType <em>Type</em>}' attribute.
	 * <!-- begin-user-doc --> <!-- end-user-doc -->
	 * @param value the new value of the '<em>Type</em>' attribute.
	 * @see org.eclipse.set.model.siteplan.PZBType
	 * @see #getType()
	 * @generated
	 */
	void setType(PZBType value);

	/**
	 * Returns the value of the '<em><b>Element</b></em>' attribute.
	 * The literals are from the enumeration {@link org.eclipse.set.model.siteplan.PZBElement}.
	 * <!-- begin-user-doc
	 * --> <!-- end-user-doc -->
	 * @return the value of the '<em>Element</em>' attribute.
	 * @see org.eclipse.set.model.siteplan.PZBElement
	 * @see #setElement(PZBElement)
	 * @see org.eclipse.set.model.siteplan.SiteplanPackage#getPZB_Element()
	 * @model
	 * @generated
	 */
	PZBElement getElement();

	/**
	 * Sets the value of the '{@link org.eclipse.set.model.siteplan.PZB#getElement <em>Element</em>}' attribute.
	 * <!-- begin-user-doc --> <!-- end-user-doc -->
	 * @param value the new value of the '<em>Element</em>' attribute.
	 * @see org.eclipse.set.model.siteplan.PZBElement
	 * @see #getElement()
	 * @generated
	 */
	void setElement(PZBElement value);

	/**
	 * Returns the value of the '<em><b>Right Side</b></em>' attribute. <!--
	 * begin-user-doc --> <!-- end-user-doc -->
	 * 
	 * @return the value of the '<em>Right Side</em>' attribute.
	 * @see #setRightSide(boolean)
	 * @see org.eclipse.set.model.siteplan.SiteplanPackage#getPZB_RightSide()
	 * @model
	 * @generated
	 */
	boolean isRightSide();

	/**
	 * Sets the value of the '{@link org.eclipse.set.model.siteplan.PZB#isRightSide <em>Right Side</em>}' attribute.
	 * <!-- begin-user-doc --> <!-- end-user-doc -->
	 * @param value the new value of the '<em>Right Side</em>' attribute.
	 * @see #isRightSide()
	 * @generated
	 */
	void setRightSide(boolean value);

	/**
	 * Returns the value of the '<em><b>Effectivity</b></em>' attribute. The
	 * literals are from the enumeration
	 * {@link org.eclipse.set.model.siteplan.PZBEffectivity}. <!--
	 * begin-user-doc --> <!-- end-user-doc -->
	 * 
	 * @return the value of the '<em>Effectivity</em>' attribute.
	 * @see org.eclipse.set.model.siteplan.PZBEffectivity
	 * @see #setEffectivity(PZBEffectivity)
	 * @see org.eclipse.set.model.siteplan.SiteplanPackage#getPZB_Effectivity()
	 * @model
	 * @generated
	 */
	PZBEffectivity getEffectivity();

	/**
	 * Sets the value of the '{@link org.eclipse.set.model.siteplan.PZB#getEffectivity <em>Effectivity</em>}' attribute.
	 * <!-- begin-user-doc --> <!--
	 * end-user-doc -->
	 * @param value the new value of the '<em>Effectivity</em>' attribute.
	 * @see org.eclipse.set.model.siteplan.PZBEffectivity
	 * @see #getEffectivity()
	 * @generated
	 */
	void setEffectivity(PZBEffectivity value);

} // PZB
